### ISSUE TITLE
[REVIEW]: robustHD: An R package for robust regression with high-dimensional data

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#3786]
**Submitting author:** @aalfons (<a href="http://orcid.org/0000-0002-2513-3788">Andreas Alfons</a>)
**Repository:** <a href="https://github.com/aalfons/robustHD/" target ="_blank">https://github.com/aalfons/robustHD/</a>
**Version:** v0.7.1
**Editor:** @mikldk
**Reviewer:** @valentint, @msalibian
**Archive:** <a href="https://doi.org/10.25397/eur.16802596.v1" target="_blank">10.25397/eur.16802596.v1</a>

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

## Status

[![status](https://joss.theoj.org/papers/3929d5aaa6df61cea6e470253df4a258/status.svg)](https://joss.theoj.org/papers/3929d5aaa6df61cea6e470253df4a258)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/3929d5aaa6df61cea6e470253df4a258"><img src="https://joss.theoj.org/papers/3929d5aaa6df61cea6e470253df4a258/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/3929d5aaa6df61cea6e470253df4a258/status.svg)](https://joss.theoj.org/papers/3929d5aaa6df61cea6e470253df4a258)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/aalfons/robustHD/" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@valentint & @msalibian, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @mikldk know.

✨ **Please start on your review when you are able, and be sure to complete your review in the next six weeks, at the very latest** ✨ 

## Review checklist for @valentint
✨ Important: Please do not use the _Convert to issue_ functionality when working through this checklist, instead, please open any new issues associated with your review <a target="_blank" href="https://github.com/aalfons/robustHD/">in the software repository associated with the submission</a>. ✨

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/aalfons/robustHD/">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@aalfons) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

## Review checklist for @msalibian
✨ Important: Please do not use the _Convert to issue_ functionality when working through this checklist, instead, please open any new issues associated with your review <a target="_blank" href="https://github.com/aalfons/robustHD/">in the software repository associated with the submission</a>. ✨

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/aalfons/robustHD/">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@aalfons) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @valentint, @msalibian it looks like you're currently assigned to review this paper :tada:.

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
Wordcount for `paper.md` is 1546

**Comment by whedon** [comment]
```
Software report (experimental):

github.com/AlDanial/cloc v 1.88  T=0.08 s (1004.9 files/s, 158530.3 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
R                               56            583           3734           3863
C++                              8            139            699           1261
SVG                              2              0              0           1053
Markdown                         2             86              0            460
TeX                              1             14              0            137
C/C++ Header                     8             46             60             95
C                                1              3              1             40
Rmd                              1             67             85             37
-------------------------------------------------------------------------------
SUM:                            79            938           4579           6946
-------------------------------------------------------------------------------


Statistical information for the repository '97cbb9e386fb99318e2109db' was
gathered on 2021/10/01.
The following historical commit information, by author, was found:

Author                     Commits    Insertions      Deletions    % of changes
Andreas Alfons                  45          4408           2075          100.00

Below are the number of rows from each author that have survived and are still
intact in the current revision:

Author                     Rows      Stability          Age       % in comments
Andreas Alfons             2344           53.2         92.2               34.56
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03786/joss.03786/10.21105.joss.03786.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03786/joss.03786/10.21105.joss.03786.pdf) :page_facing_up: :point_left:

**Comment by mikldk** [comment]
@valentint, @msalibian: Thanks for agreeing to review. Please carry out your review in this issue by updating the checklist above and giving feedback in this issue. The reviewer guidelines are available here: <https://joss.readthedocs.io/en/latest/reviewer_guidelines.html>. If possible create issues (and cross-reference) in the submission's repository to avoid too specific discussions in this review thread. 

If you have any questions or concerns please let me know.

Please note that I asked about the length of the paper in the [pre-review](https://github.com/openjournals/joss-reviews/issues/3779#issuecomment-931275964). Both @aalfons and I would like your input in that. Bear in mind the JOSS guidelines of what the paper should contain: <https://joss.readthedocs.io/en/latest/submitting.html#what-should-my-paper-contain>.

**Comment by msalibian** [comment]
@mikldk I'm happy with the submission (documentation & repository are good to go, as far as I can see). For the paper I have a few suggestions below. 

The issue of length is a tricky one: yes, the paper is long, but the two examples illustrate different functionality of the software. One way around this could be to reduce the text output of the examples in the paper (e.g. remove the estimated regression coefficients, etc.) and shorten the description of the data, etc. I think Figure 1 may also go, but I'd keep Figure 2 (the main difference between them being that in Fig 1 one does not get the whole 'path of solutions', because either `sparseLTS` doesn't compute it, or because it's not computed when using prediction error (as opposed to BIC) to select the optimal model, which is something that I'd like @aalfons to clarify, see my comment below). 

Specific suggestions: 
- Summary: I'd avoid saying "cleaning the data" (lines 18-19 of page 1), specially when advocating for the use of robust estimation methods. Such a phrase opens the door to considering cleaning the data and applying "classical" methods as an alternative "robust" methodology, which is well-known to not work well in general;
- Second paragraph of "Statement of need", rather than referring to the historical value of the package, I'd focus on what it does: implements regularized trimmed-squares regression estimators, the plug-in LARS and its group-wise variant;
- line 57 (page 2), does `sparseLTS` **estimate** the smallest lambda that sets all coefficients to zero, or does it **compute** it? 
- line 58, what other options are available for the arguments `mode` and `crit`?
- line 60, presumably the cross-validated prediction error also uses some kind of trimming or robustification, this should be mentioned when discussing how the prediction error is estimated;
- line 66, although the argument `K` of `foldControl` seems fairly self explanatory, I think it'd be helpful to mention what `R` is or does;
- lines 68-79, it'd also be helpful to say something about what is in the "reweighted" column (as compared with the "raw" one);
- line 104, comparing this with the robust lars example, can we also get a full path of solutions with `sparseLTS` if we use `crit = "BIC"`? the different default behaviour is a bit intriguing;
- lines 130-137: how are groups formed? It does not seem to be specified explicitly in the call to `rgrplars` below; is there a default behaviour? can the user override it? 
- lines 135-137: I don't think I'm understanding this sentence correctly: "The robust regression estimator used to estimate the submodels uses an initial subsampling of observations(...)" Should "estimate the submodels" be "estimate the regression coefficients corresponding to each submodel" (or something along those lines)? Or is it really trying to determine which groups of features (and thus submodels) to use? 
- Also regarding the sentence just above: I seem to remember that the "original" Robust LARS simply run the classical LARS algorithm using robust correlation coefficients, and thus it did not need iterations or initial values, etc. Am I remembering it wrong?

**Comment by aalfons** [comment]
@mikldk @msalibian @valentint - I created a revised version of the manuscript which is reduced in length and fits on 5 pages. I'll give detailed responses to each reviewer in separate comments below.

**Comment by aalfons** [comment]
Reply to @msalibian:

Thanks again for your helpful suggestions on reducing the length. I decided to keep the output on the regression coefficients, but for `sparseLTS()` I excluded the output regarding the cross-validation.  This also avoided a discussion on the raw and reweighted estimators (see also my reply to one of the comments below). Moreover, I followed your suggestion and removed Figure 1 but kept Figure 2.  With this change, it seemed more appropriate to present the `rgrplars()` example first, as it is more detailed.  Furthermore, I think that in this case the length of the paper can be further reduced by avoiding the discussion on when the coefficient plot is available (see my reply to one of the comments below for details). 

ad 1) (avoid saying "cleaning the data") -> Done (see page 1, line 19).

ad 2) (shortening the Statement of Need) -> As suggested, I removed two sentences on the historical value of the package, but I kept the sentence referring to the use of `robustHD` in the literature, as it underlines its usefulness (page 1, lines 29-32). But I'd be happy for some direction by @mikldk on this (see also my reply below to a related point by @valentint).

ad 3) (**estimate** or **compute** smallest lambda that sets everything to zero) -> As clarified this in my earlier reply, the word "estimate" is correct here.

ad 4) (other options for the arguments `mode` and `crit`) -> In the spirit of not (unnecessarily) extending the length of the paper, I think it is better not to include a discussion on other options for the arguments `mode` and `crit`.  Interested readers can find this information in the help file.

ad 5) (cross-validated prediction error) -> I added a sentence that for sparse LTS, the default prediction loss function
is the root trimmed mean squared prediction error (page 3, lines 114-115).

ad 6) (clarification of arguments of `foldControl()`) -> I took greater care to explain the different arguments in the function call, including an explanation that `splits = foldControl(K = 5, R = 1)` defines 5-fold cross-validation with one replication (page 3, lines 112-114).

ad 7) (clarifying "reweighted" and "raw" columns of output) -> To shorten the paper, I followed your suggestion to remove some of the text output in the `sparseLTS()` example.  The cross-validation results are no longer shown, avoiding a discussion on the "raw" and "reweighted" columns altogether.

ad 8) (whether we can get a full path of `sparseLTS()` solutions with `crit = "BIC"`) -> As clarified in my earlier reply, this behavior has to do with how the final model is selected, irrespective of the modeling technique.  With switching the order of the examples and removing the plots for the `sparseLTS()` example (see the discussion above), I found it better to avoid this discussion altogether to shorten the paper. Regarding plots of results from `sparseLTS()`, I now simply write: "Similar plots as in Figure 1 are available to visualize the results." (page 3, lines 140-141)  This is a bit of an oversimplification, as the coefficient plot is not available for this specific example, but I think that this simplification acceptable since the plot is indeed available if `crit = "BIC"` is used in `sparseLTS()`.

ad 9) (how groups are formed in `rgrplars()`) -> I rewrote the corresponding paragraph.  Specifically, I now write: "Through the formula interface, function `rgrplars()` by default takes each categorical variable (`factor`) as a group of dummy variables while all remaining variables are taken individually.  However, the group assignment can be defined by the user through argument `assign`." (page 2, lines 53-56)

ad 10) (confusing sentence regarding the robust regression estimator in `rgrplars()`) -> In line with the clarification from my earlier reply, I now write: "Note that each submodel along the sequence is fitted using a robust regression estimator with a non-deterministic algorithm, hence the seed of the random number generator is supplied for reproducibility." (page 2, lines 58-60)

ad 11) (clarification on robust LARS) -> As already clarified in my earlier reply, also robust LARS uses a two-step strategy where first the variables are sequences, and then submodel along the sequence are estimated with robust regressions.

Thanks again for the helpful comments, and please let me know if anything is still unclear.

**Comment by valentint** [comment]
@mikldk First of all I would like to recommend the revised paper for publishing.
The technical issues I had mentioned before could be discussed out of this review.

@aalfons Installation: yes, it seems that it is my local problem, but it occurs on all my PCs :) and I still cannot find a way out of it. Since I have had the issue with other R packages, it is clear that it does not concern robustHD.

@aalfons @msalibian The timing: I got curious now :) I am on a relatively new Surface Book, i7 processor, and 8GB RAM. Please find below the script with my approximate timings and let me know if I am doing something wrong.

library("robustHD")
library(tictoc)

data("nci60")

y <- protein[, 92]
correlations <- apply(gene, 2, corHuber, y)
keep <- partialOrder(abs(correlations), 100, decreasing = TRUE)
X <- gene[, keep]

##  lambda <- seq(0.01, 0.5, length.out=10)

## First I run sparseLTS() with the  sequence of 10 lamdas, as in the example on GitHub - it takes approx 5 minutes.
lambda <- seq(0.01, 0.5, length.out=10)
tic()
fit <- sparseLTS(X, y, lambda=lambda, mode = "fraction", crit = "PE", splits = foldControl(K = 5, R = 1), seed = 20210507)
toc()

##  319.12 sec elapsed

## 2. Then I run three times sparseLTS() for three different values of lambda - it is quite fast, less then 10 seconds each
tic()
fit <- sparseLTS(X, y, lambda=0.01, mode = "fraction", crit = "PE", splits = foldControl(K = 5, R = 1), seed = 20210507)
fit <- sparseLTS(X, y, lambda=0.255, mode = "fraction", crit = "PE", splits = foldControl(K = 5, R = 1), seed = 20210507)
fit <- sparseLTS(X, y, lambda=0.05, mode = "fraction", crit = "PE", splits = foldControl(K = 5, R = 1), seed = 20210507)
toc()

##  28.01 sec elapsed

## 2. And finally I run sparseLTS() with a grid containing the same three lambdas - the elapsed time is more than five times larger
lambda <- c(0.01, 0.255, 0.5)
tic()
fit <- sparseLTS(X, y, lambda=lambda, mode = "fraction", crit = "PE", splits = foldControl(K = 5, R = 1), seed = 20210507)
toc()

##  186.92 sec elapsed

**Comment by msalibian** [comment]
@valentint This is intriguing. Below I copy my Surface system information, and here's the output of the first part of your script. I don't know what may explain the difference. 

```
> library("robustHD")
Loading required package: ggplot2
Loading required package: perry
Loading required package: parallel
Loading required package: robustbase
Warning messages:
1: package ‘perry’ was built under R version 4.1.1 
2: package ‘robustbase’ was built under R version 4.1.1 
> library(tictoc)
Warning message:
package ‘tictoc’ was built under R version 4.1.1 
> data("nci60")
> y <- protein[, 92]
> correlations <- apply(gene, 2, corHuber, y)
> keep <- partialOrder(abs(correlations), 100, decreasing = TRUE)
> X <- gene[, keep]
> lambda <- seq(0.01, 0.5, length.out=10)
> tic()
> fit <- sparseLTS(X, y, lambda=lambda, mode = "fraction", crit = "PE", splits = foldControl(K = 5, R = 1), seed = 20210507)
> toc()
28.5 sec elapsed
> 
> system.time(
+   fit <- sparseLTS(X, y, lambda=lambda, mode = "fraction", crit = "PE", splits = foldControl(K = 5, R = 1), seed = 20210507)
+ )
   user  system elapsed 
  26.91    0.01   27.98 
> sessionInfo()
R version 4.1.0 (2021-05-18)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 19043)

Matrix products: default

locale:
[1] LC_COLLATE=English_Canada.1252  LC_CTYPE=English_Canada.1252   
[3] LC_MONETARY=English_Canada.1252 LC_NUMERIC=C                   
[5] LC_TIME=English_Canada.1252    

attached base packages:
[1] parallel  stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] tictoc_1.0.1      robustHD_0.7.1    robustbase_0.93-9 perry_0.3.0       ggplot2_3.3.5    

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.7       magrittr_2.0.1   MASS_7.3-54      tidyselect_1.1.1 munsell_0.5.0   
 [6] colorspace_2.0-2 R6_2.5.1         rlang_0.4.11     fansi_0.5.0      dplyr_1.0.7     
[11] tools_4.1.0      grid_4.1.0       gtable_0.3.0     utf8_1.2.2       withr_2.4.2     
[16] ellipsis_0.3.2   tibble_3.1.5     lifecycle_1.0.1  crayon_1.4.1     purrr_0.3.4     
[21] vctrs_0.3.8      glue_1.4.2       compiler_4.1.0   DEoptimR_1.0-9   pillar_1.6.3    
[26] generics_0.1.0   scales_1.1.1     pkgconfig_2.0.3 
```



![image](https://user-images.githubusercontent.com/5422467/137037223-ed340322-5071-4082-9cd1-52e60ed245e6.png)

**Comment by mikldk** [comment]
@aalfons Maybe it is due to underlying BLAS/LAPACK? Anyway, please let me know once the software and paper is ready (you can better judge which of the above issues are resolved or not). Once software and paper is ready, please:

* Please have a final read though of the paper, checking language etc.
* Have a final check of the proofs with `@whedon generate pdf`
* In general, be sure that all versions are correct and that the repo is in the state that should be published.
* Please make a tagged release and archive (e.g. with [Zenodo](https://zenodo.org/)) as described [here](<https://joss.readthedocs.io/en/latest/submitting.html#the-review-process>), and report the version number and archive DOI in this thread. Please verify that the archive deposit has the correct metadata (title and author list), or edit these if that is not the case.

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03786/joss.03786/10.21105.joss.03786.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03786/joss.03786/10.21105.joss.03786.pdf) :page_facing_up: :point_left:

**Comment by aalfons** [comment]
@kthyng - The metadata on the archive is now publicly visible, including the sentence regarding the version number.

It turns out that the change in the metadata didn't trigger a new DOI after all, the current DOI is still correct. So everything should now be ready to go.

**Comment by kthyng** [comment]
great!

**Comment by kthyng** [comment]
@whedon accept deposit=true

**Comment by whedon** [comment]
```
Doing it live! Attempting automated processing of paper acceptance...
```

**Comment by whedon** [comment]
🐦🐦🐦 👉 [Tweet for this paper](https://twitter.com/JOSS_TheOJ/status/1455944629254246401) 👈 🐦🐦🐦

**Comment by whedon** [comment]
🚨🚨🚨 **THIS IS NOT A DRILL, YOU HAVE JUST ACCEPTED A PAPER INTO JOSS!** 🚨🚨🚨

 Here's what you must now do:

0. Check final PDF and Crossref metadata that was deposited :point_right: https://github.com/openjournals/joss-papers/pull/2727
1. Wait a couple of minutes, then verify that the paper DOI resolves [https://doi.org/10.21105/joss.03786](https://doi.org/10.21105/joss.03786)
2. If everything looks good, then close this review issue.
3. Party like you just published a paper! 🎉🌈🦄💃👻🤘

 Any issues? Notify your editorial technical team...

**Comment by kthyng** [comment]
Congrats on your new publication @aalfons! Many thanks to editor @mikldk and to reviewers @valentint and @msalibian for your time, hard work, and expertise!!

(I will leave this issue open until the DOI resolves)

**Comment by aalfons** [comment]
Thanks @kthyng - the DOI seems to work!

And thanks again @mikldk, @valentint, and @msalibian for the constructive comments.

**Comment by danielskatz** [comment]
(closing, as the DOI works for me as well)

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](https://joss.theoj.org/papers/10.21105/joss.03786/status.svg)](https://doi.org/10.21105/joss.03786)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/joss.03786">
  <img src="https://joss.theoj.org/papers/10.21105/joss.03786/status.svg" alt="DOI badge" >
</a>

reStructuredText:
.. image:: https://joss.theoj.org/papers/10.21105/joss.03786/status.svg
   :target: https://doi.org/10.21105/joss.03786
```
This is how it will look in your documentation:

[![DOI](https://joss.theoj.org/papers/10.21105/joss.03786/status.svg)](https://doi.org/10.21105/joss.03786)

**We need your help!**

Journal of Open Source Software is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: https://joss.theoj.org/reviewer-signup.html
- Making a small donation to support our running costs here: https://numfocus.org/donate-to-joss
